File: lib/formats/gfx/tlg_converter.rb
require 'lib/binary_io'
require 'lib/image'
require_relative 'tlg_converter/tlg5_pixel_decoder'

# Converts TLG to PNG.
# Seen in XP3 archives.
module TlgConverter
  module_function

  MAGIC_TLG0 = "TLG0.0\x00sds\x1a"
  MAGIC_TLG5 = "TLG5.0\x00raw\x1a"
  MAGIC_TLG6 = "TLG6.0\x00raw\x1a"

  def add_cli_help(_arg_parser) end

  def parse_cli_options(_arg_parser, _options) end

  def decode!(file, _options)
    Decoder.new.read(file.data).update_file(file)
  end

  def encode!(_file, _options)
    fail 'Not supported'
  end

  class Decoder
    def read(data)
      input = BinaryIO.from_string(data)
      if data.start_with?(MAGIC_TLG0)
        input.seek(MAGIC_TLG0.length)
        return read_tlg0(input)
      elsif data.start_with?(MAGIC_TLG5)
        input.seek(MAGIC_TLG5.length)
        return read_tlg5(input)
      elsif data.start_with?(MAGIC_TLG6)
        input.seek(MAGIC_TLG6.length)
        return read_tlg6(input)
      else
        fail 'Not a TLG image.'
      end
    end

    def read_tlg0(_input)
      fail 'Not supported (yet)'
    end

    def read_tlg5(input)
      header = {}
      header[:channel_count],
      header[:image_width],
      header[:image_height],
      header[:block_height] = input.read(13).unpack('CL4')
      pixels = input.read

      output = decode_tlg5_pixels(header, pixels)
      Image.from_pixels(
        header[:image_width],
        header[:image_height],
        output,
        'RGBA')
    end

    def read_tlg6(_input)
      fail 'Not supported (yet)'
    end
  end
end

File: src/formats/gfx/tlg_converter.cc
// TLG image
//
// Company:   -
// Engine:    Kirikiri
// Extension: .tlg
// Archives:  XP3
//
// Known Games:
// - Fate/Stay Night
// - Fate/Hollow Ataraxia

#include <stdexcept>
#include "formats/gfx/tlg_converter.h"
#include "formats/gfx/tlg_converter/tlg5_decoder.h"
#include "formats/gfx/tlg_converter/tlg6_decoder.h"

namespace
{
    const std::string magic_tlg_0("TLG0.0\x00sds\x1a", 11);
    const std::string magic_tlg_5("TLG5.0\x00raw\x1a", 11);
    const std::string magic_tlg_6("TLG6.0\x00raw\x1a", 11);

    void guess_version_and_decode(File &file);

    std::string extract_string(std::string &container)
    {
        size_t length = 0;
        while (container.length() && container[0] >= '0' && container[0] <= '9')
        {
            length *= 10;
            length += container[0] - '0';
            container.erase(0, 1);
        }
        container.erase(0, 1);
        std::string str = container.substr(0, length);
        container.erase(0, length + 1);
        return str;
    }

    void decode_tlg_0(File &file)
    {
        size_t raw_data_size = file.io.read_u32_le();
        size_t raw_data_offset = file.io.tell();

        std::vector<std::pair<std::string, std::string>> tags;

        file.io.skip(raw_data_size);
        while (file.io.tell() < file.io.size())
        {
            std::string chunk_name = file.io.read(4);
            size_t chunk_size = file.io.read_u32_le();
            std::string chunk_data = file.io.read(chunk_size);

            if (chunk_name == "tags")
            {
                while (chunk_data != "")
                {
                    std::string key = extract_string(chunk_data);
                    std::string value = extract_string(chunk_data);
                    tags.push_back(
                        std::pair<std::string, std::string>(key, value));
                }
            }
            else
                throw std::runtime_error("Unknown chunk: " + chunk_name);
        }

        file.io.seek(raw_data_offset);
        guess_version_and_decode(file);
    }

    void decode_tlg_5(File &file)
    {
        Tlg5Decoder decoder;
        decoder.decode(file);
    }

    void decode_tlg_6(File &file)
    {
        Tlg6Decoder decoder;
        decoder.decode(file);
    }

    void guess_version_and_decode(File &file)
    {
        size_t pos = file.io.tell();

        file.io.seek(pos);
        if (file.io.read(magic_tlg_0.size()) ==  magic_tlg_0)
        {
            decode_tlg_0(file);
            return;
        }

        file.io.seek(pos);
        if (file.io.read(magic_tlg_5.size()) ==  magic_tlg_5)
        {
            decode_tlg_5(file);
            return;
        }

        file.io.seek(pos);
        if (file.io.read(magic_tlg_6.size()) ==  magic_tlg_6)
        {
            decode_tlg_6(file);
            return;
        }

        throw std::runtime_error("Not a TLG image");
    }
}

void TlgConverter::decode_internal(File &file) const
{
    guess_version_and_decode(file);
}
